### PR TITLE
dird: cats: adapt `purge` command to delete jobs with specific jobstatus and/or from specific pool

### PR DESCRIPTION
**Pull request by alaaeddineelamri**
#### Description

This PR adds the ability for the `purge` command to delete jobs with a specific jobstatus and/or from specific pool.
The command used to only do so for specific volumes or clients.

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)

##### General
- [x] PR name is meaningful
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Check backport line
- [x] Is the PR title usable as CHANGELOG entry?
- ~~Separate commit for CHANGELOG.md ("update CHANGELOG.md"). The PR number is correct.~~

##### Source code quality

- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR
- [x] `bareos-check-sources --since-merge` does not report any problems

##### Tests

- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

File: core/src/dird/ua_purge.cc [hunk 1]
```diff
@@ -262,47 +278,93 @@ static bool PurgeFilesFromClient(UaContext* ua, ClientResource* client)
...
+        = CreateDelimitedStringForSqlQueries(jobstatuslist, ',');
+    select_jobs_from_pool += "AND Job.JobStatus IN (" + jobStatuses + ") ";
+  }
+  select_jobs_from_pool.append("ORDER BY JobId");
+
+
+  std::vector<JobId_t> delete_list;
+  ua->db->SqlQuery(select_jobs_from_pool.c_str(), JobDeleteHandler,
+                   &delete_list);
+
```
Comment (joergsteffens) [inline core/src/dird/ua_purge.cc:307]: PurgeJobsFromClient asks for confirmation before purging. Shouldn'd PurgeJobsFromPool do the same? (PurgeJobsFromVolume also does not ask).

File: core/src/dird/ua_purge.cc [hunk 2]
```diff
@@ -262,47 +278,93 @@ static bool PurgeFilesFromClient(UaContext* ua, ClientResource* client)
...
+  if (delete_list.size() > 0) { PurgeJobListFromCatalog(ua, delete_list); }
+
```
Comment (joergsteffens) [inline core/src/dird/ua_purge.cc:309]: When  purging all jobs from a volume, the volume is still marked as full. When purging the last job from a volume, the volume will be marked as purged afterwards.

**Review by joergsteffens — CHANGES_REQUESTED** [review]

**Comment by joergsteffens** [comment]
I also did some modifications directly in commits. Unfortunately, I'd not have the permission to modify this PR directly (I though, did is the default?). Instead I pushed my suggestions to https://github.com/joergsteffens/bareos/tree/dev/joergs/master/s5201-purge-failed-jobs

File: core/src/dird/ua_purge.cc [hunk 3]
```diff
@@ -467,28 +543,25 @@ bool PurgeJobsFromVolume(UaContext* ua, MediaDbRecord* mr, bool force)
...
-       edit_int64(mr->MediaId, ed1));
+  Mmsg(query, "SELECT 1 FROM JobMedia WHERE MediaId=%lu LIMIT 1", mr->MediaId);
   if (!ua->db->SqlQuery(query.c_str(), DelCountHandler,
                         static_cast<void*>(&cnt))) {
     ua->ErrorMsg("%s", ua->db->strerror());
     Dmsg0(050, "Count failed\n");
     return false;
   }
 
+  bool purged = false;
```
[the remark this answers is not in the document]
Comment (alaaeddineelamri) [reply core/src/dird/ua_purge.cc:564]: it's still used just underneath

File: systemtests/tests/pruning/testrunner [hunk 4]
```diff
@@ -149,87 +166,136 @@ prune files yes
...
 
-# Prune all volumes of specific pool
+cat <<END_OF_DATA >"$tmp/bconcmds"
+@$out $purging_failed_jobs
+run job=$failingjob_name level=Full yes
+run job=$failingjob_name level=Full yes
+run job=$failingjob_name level=Full yes
+wait
+messages
+purge jobs client jobstatus=E yes
```
Comment (joergsteffens) [inline systemtests/tests/pruning/testrunner:186]:
```suggestion
purge jobs client=bareos-fd jobstatus=E yes
```
Cosmetical change, but without I've been confused at first (if only only client is configured, it gets autoselected).

**Review by joergsteffens — CHANGES_REQUESTED** [review]
Nearly done, but pr-tool complains:
```
 ✗  Commit checks failed:
	f874336d4 refactor: systemtest: updated systemtest: body contains line longer 72 chars
	aa771a59e refactor: ua_purge.cc: remove bailouts, rename and reorder variables: headline too long
 ✗  bareos-check-sources --since=421f8eb125967d7faac63cf9acf0eddff16d2244 reported:
	Plugin 'compress c block comments' would modify 'core/src/cats/sql.cc'
	Plugin 'fix copyright year' would modify 'core/src/cats/sql.cc'
	Plugin 'fix copyright year' would modify 'core/src/cats/sql.h'
```

**Review by joergsteffens — APPROVED** [review]

**Pull request merged**